app/src/main/java/com/example/livedata/LoginFragment.kt

```diff
@@ -0,0 +1,58 @@
+package com.example.livedata
+
+import android.arch.lifecycle.Observer
+import android.arch.lifecycle.ViewModelProviders
+import android.databinding.DataBindingUtil
+import android.os.Bundle
+import android.support.v4.app.Fragment
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.EditText
+import com.example.livedata.databinding.FragmentLoginBinding
+
+
+class LoginFragment : Fragment() {
+    private lateinit var loginViewModel: LoginViewModel
+    private lateinit var binding: FragmentLoginBinding
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        super.onCreate(savedInstanceState)
+        loginViewModel = ViewModelProviders.of(this).get(LoginViewModel::class.java)
+        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_login, container, false)
+        binding.let {
+            it.lifecycleOwner = this
+            it.loginViewModel = loginViewModel
+            return it.root
+        }
```

Comment:
Since fragment lifecycles are a bit more complicated than activity's lifecycle (and causes issues, you can read about it [here](https://medium.com/@BladeCoder/architecture-components-pitfalls-part-1-9300dd969808)), they added a lifecycle that represents the fragment's. Use ```getViewLifecycleOwner()``` instead.

I think it'd be cleaner if you separate the field initialization and the return lines like:
```
return binding.apply {
    lifecycleOwner = getViewLifecycleOwner()
    loginViewModel = loginViewModel
}.root
```
But not necessary

Reply:
This does not work without having to name the private variable loginViewModel something different. apply changes scope to this so any reference to loginViewModel will imply this. loginViewModel = this.loginViewModel

Reply:
when you use let it.loginViewModel is the attribute on binding and loginViewModel is the private variable in the Fragment